Mathlib/Tactic/Linter/EmptyLine.lean

```diff
@@ -0,0 +1,152 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Mathlib.Tactic.Linter.Header
+
+/-!
+#  The "emptyLine" linter
+
+The "emptyLine" linter emits a warning on empty lines inside a command, but outside of a
+doc-string/module-doc.
+-/
+
+open Lean Elab
+
+/-- Retrieve the `String.Range` of a `Substring`. -/
+def Lean.Substring.getRange : Substring → String.Range
+  | {startPos := st, stopPos := en, ..} => ⟨st, en⟩
+
+namespace Lean.Syntax
+/-!
+# `Syntax` filters
+-/
+
+/--
+`filterMapM stx f` takes as input a `Syntax` `stx` and a monadic function `f : Syntax → m α`.
+It produces the array of the `some` values that `f` takes while traversing `stx`.
+
+See `filterMap` for a non-monadic version.
+-/
+partial
+def filterMapM {m : Type → Type} [Monad m] {α} (stx : Syntax) (f : Syntax → m (Option α)) :
+    m (Array α) := do
+  let nargs := (← stx.getArgs.mapM (·.filterMapM f)).flatten
+  match ← f stx with
+    | some new => return nargs.push new
+    | none => return nargs
+
+/--
+`filterMap stx f` takes as input a `Syntax` `stx` and a function `f : Syntax → α`.
+It produces the array of the `some` values that `f` takes while traversing `stx`.
+
+See `filterMapM` for a non-monadic version.
+-/
+def filterMap {α} (stx : Syntax) (f : Syntax → Option α) : Array α :=
+  stx.filterMapM (m := Id) f
+
+/--
+`filter stx f` takes as input a `Syntax` `stx` and a function `f : Syntax → Bool`.
+It produces the array of the syntax nodes in `stx` where `f` is `true`.
+
+See also the similar `filterMap` and `filterMapM`.
+-/
+def filter (stx : Syntax) (f : Syntax → Bool) : Array Syntax :=
+  stx.filterMap fun s => if f s then some s else none
+
+end Lean.Syntax
+
+namespace Mathlib.Linter
+
+/--
+The "emptyLine" linter emits a warning on empty lines inside a command, but outside of a
+doc-string/module-doc.
+
+The linter is only active when there are no other warning, so as to not add noise when developing
+incomplete proofs.
+-/
+register_option linter.style.emptyLine : Bool := {
+  defValue := false
+  descr := "enable the emptyLine linter"
+}
+
+namespace EmptyLine
+
+/-- The `SyntaxNodeKind`s where the `EmptyLine` linter is not active. -/
+abbrev AllowEmptyLines : Std.HashSet SyntaxNodeKind := Std.HashSet.emptyWithCapacity
+  |>.insert ``Parser.Command.docComment
+  |>.insert ``Parser.Command.moduleDoc
+  |>.insert ``Parser.Command.mutual
+  |>.insert `str
+
+/-- If a file contains one of these names as segments, we disable the `emptyLine` linter. -/
+abbrev SkippedFileSegments : Std.HashSet Name := Std.HashSet.emptyWithCapacity
+  |>.insert `Tactic
```

Comment:
Add a test for one of these?

Reply:
This is possible, but would mean adding a new file whose path matches one of these segments. I don't think this is necessarily worth it, and am fine landing this without.

Reply:
I think that this is not really worth it: there are files with `Tactic` in their name, so, in some sense, those files serve as tests already.

Besides, maybe in the future the linter should expand to look at "meta" files, maybe with some more refined heuristic than the current one.